openapi_python_client/parser/properties/__init__.py

```diff
@@ -167,18 +159,37 @@ def __attrs_post_init__(self) -> None:
             self, "has_properties_without_templates", any(prop.template is None for prop in self.inner_properties)
         )
 
-    def get_type_string(self, no_optional: bool = False) -> str:
-        """ Get a string representation of type that should be used when declaring this property """
+    def _get_inner_prop_string(self) -> str:
         inner_types = [p.get_type_string(no_optional=True) for p in self.inner_properties]
-        inner_prop_string = ", ".join(inner_types)
-        type_string = f"Union[{inner_prop_string}]"
+        return ", ".join(inner_types)
+
+    def get_base_type_string(self) -> str:
+        return f"Union[{self._get_inner_prop_string()}]"
+
+    def get_type_string(self, no_optional: bool = False, query_parameter: bool = False) -> str:
+        """
+        Get a string representation of type that should be used when declaring this property.
+
+        This implementation differs slightly from `Property.get_type_string` in order to collapse
+        nested union types.
+        """
+        type_string = self.get_base_type_string()
         if no_optional:
             return type_string
-        if not self.required:
-            type_string = f"Union[Unset, {inner_prop_string}]"
-        if self.nullable:
-            type_string = f"Optional[{type_string}]"
-        return type_string
+        if self.required:
+            if self.nullable:
+                return f"Union[None, {self._get_inner_prop_string()}]"
+            else:
+                return type_string
+        else:
+            if self.nullable:
+                return f"Union[Unset, None, {self._get_inner_prop_string()}]"
+            else:
+                if query_parameter:
+                    # For query parameters, None has the same meaning as Unset
+                    return f"Union[Unset, None, {self._get_inner_prop_string()}]"
+                else:
+                    return f"Union[Unset, {self._get_inner_prop_string()}]"
```

Comment:
I think this cleans up the logic a little bit (one case per output option) and covers the case of `required and nullable and query_parameter` correctly. Spacing may be wrong because I'm editing in GitHub >.>.

```suggestion
        if self.required and not self.nullable:
            return type_string
        if query_parameter or (self.nullable and not self.required):
            # For query parameters, nullable and required are the same
            return f"Union[Unset, None, {self._get_inner_prop_string()}]"
        if self.nullable:
            return f"Union[None, {self._get_inner_prop_string()}]"
        return f"Union[Unset, {self._get_inner_prop_string()}]"
```

Reply:
I think the case that differs is if `query_parameter=True, required=True, nullable=True`. In the original code, we get `f"Union[None, {self._get_inner_prop_string()}]"`, whereas in the suggestion, we get `f"Union[Unset, None, {self._get_inner_prop_string()}]"`. It seems like the original code would be correct in that case, unless it's the case that query parameters cannot exist with `required=True, nullable=True`?

Reply:
Yeah, hmm, it seems like the semantics are weird for a query parameter with `required=True, nullable=True`...what would it mean for you to pass `None`?

This also impacts your other comment in `endpoint_macros.pyi`...if we decide that we aren't gonna allow passing `None`, then I think we also need `{% if property.required and not property.nullable %}` in L25.

This is so brain-twisting 🤔

Reply:
@dbanty (cc @bowenwr) OK I think I figured out why I'm so confused. Seems like for query parameters, it's redundant to have BOTH `required` and `nullable` as properties, because there's no distinction between `None` and `UNSET`. 

In the current implementation, we have:
![image](https://user-images.githubusercontent.com/64445763/105103039-dc65c100-5a7d-11eb-902e-a7dde90f2e03.png)

However, it's confusing because in the last row, you specifically said `nullable=False`, but still it's allowing `None`.

Is it possible to ban the `required` attribute on query parameters? That would simplify things a lot. Though it would be a breaking change.

Reply:
I think what we're going for here is to interpret both nullable and required to mean the same thing in the context of query parameters. Basically, there is no standard way to pass `null` through a query parameter, so rather than setting our own decision we're opting to not have a `null` value. In which case, `null` and `not required` become the same thing semantically when using this generator, and therefore should be the same signature.

Interestingly, in my response to #295 I encountered the "style" attribute which does seem to indicate that an "empty" value should be provided as empty, as the generated clients do today. It's not clear if empty is equivalent to null though...

Reply:
I see. In that case @dbanty do you think it would be reasonable to stick to the original implementation, which converts `query_parameter=True, required=True, nullable=True` into `Union[None, {self._get_inner_prop_string()}]`?